src/nccl_ofi_net.c

```diff
@@ -487,7 +542,7 @@ static ncclResult_t register_mr_buffers(ofiComm_t *comm, void *data,
 	/* Initialize MR attributes */
 	mr_attr.mr_iov = &iov;
 	mr_attr.iov_count = 1;
-	mr_attr.access = FI_SEND | FI_RECV;
+	mr_attr.access = FI_SEND | FI_RECV | FI_REMOTE_READ;
```

Comment:
Not sure why we need `FI_REMOTE_READ` if we only use send/recv APIs. We should although add `FI_READ` for flush buffers.

Reply:
I think you either need both `FI_READ` and `FI_REMOTE_READ` or neither.  But the definition of remote endpoint is vague enough that I think you need both.  https://ofiwg.github.io/libfabric/v1.16.0/man/fi_getinfo.3.html says that `FI_READ` says that the endpoint can initiate the read and `FI_REMOTE_READ` says the endpoint / region can be the target of the read operation.  I can see providers saying that any read is remote, so I don't really have a problem with requiring both.

Reply:
But these are memory registration attributes, from https://ofiwg.github.io/libfabric/v1.16.1/man/fi_mr.3.html, since flush operations are initiated locally and are the destination buffers of RDMA read operation. Shouldn't `FI_READ` be sufficient by itself?

> FI_READ
>     The memory buffer may be used as the result buffer for RMA read and atomic operations on the initiator side. Note that from the viewpoint of the application, the memory buffer is being written into by the network.

Whereas for `FI_REMOTE_READ`, it talks about the local buffer to be the source buffer for RDMA operation.

> FI_REMOTE_READ
>    The memory buffer may be used as the source buffer of an RMA read operation on the target side. The contents of the memory buffer are not modified by such operations.

Reply:
You're right; I got confused about which use of FI_REMOTE_READ we were talking about.  But I think I'm still right that you need both if flush is implemented with a read operation.  Today, `register_mr_buffers()` is called from both `ofi_regMr()` (which is the data buffer likely in the GPU, in which case it may be the source for a flush and therefore needs `FI_REMOTE_READ` AND `alloc_and_reg_flush_buff()`, in which case the buffer is the target of the flush read and needs `FI_READ`.  So we would either need both or we need to pull apart the flags in `register_mr_buffers()` so that we knew whether this was a GPU buffer or a host buffer.  I don't think that complexity is worth the gain.

Reply:
Yes, this makes sense. One simplification could be that we add `FI_READ` to host buffers and `FI_REMOTE_READ` to GPU buffers. But, I do not feel strongly for any option so @jdinan, it would be your choice :)

Reply:
Great catch -- I missed that this change took away the `FI_READ` flag on the result buffers allocated through `alloc_and_reg_flush_buff`. It's easy to set `FI_REMOTE_READ` on CUDA buffers and `FI_READ` on host buffers.